Sources/Build/BuildPlan.swift

```diff
@@ -0,0 +1,486 @@
+        case .clang(let target):
+            return target.module
+        }
+    }
+
+    /// The objects in this target.
+    var objects: [AbsolutePath] {
+        switch self {
+        case .swift(let target):
+            return target.objects
+        case .clang(let target):
+            return target.objects
+        }
+    }
+}
+
+/// Target description for a Clang module i.e. C language family module.
+public final class ClangTargetDescription {
+
+    /// The module described by this target.
+    public let module: ClangModule
+
+    /// The build parameters.
+    let buildParameters: BuildParameters
+
+    /// The modulemap file for this target, if any.
+    private(set) var moduleMap: AbsolutePath?
+
+    /// Path to the temporary directory for this target.
+    var tempsPath: AbsolutePath {
+        return buildParameters.buildPath.appending(component: module.c99name + ".build")
+    }
+
+    /// The objects in this target.
+    var objects: [AbsolutePath] { 
+        return compilePaths().map{$0.object} 
+    }
+
+    /// Any addition flags to be added. These flags are expected to be computed during build planning.
+    fileprivate var additionalFlags: [String] = []
+    
+    /// Create a new target description with module and build parameters.
+    init(module: ClangModule, buildParameters: BuildParameters) throws {
+        self.module = module
+        self.buildParameters = buildParameters 
+        // Try computing modulemap path for a C library.
+        if module.type == .library {
+            self.moduleMap = try computeModulemapPath()
+        }
+    }
+
+    /// An array of tuple containing filename, source, object and dependency path for each of the source in this target.
+    public func compilePaths() -> [(filename: RelativePath, source: AbsolutePath, object: AbsolutePath, deps: AbsolutePath)] {
+        return module.sources.relativePaths.map { source in
+            let path = module.sources.root.appending(source)
+            let object = tempsPath.appending(RelativePath(source.asString + ".o"))
+            let deps = tempsPath.appending(RelativePath(source.asString + ".d"))
+            return (source, path, object, deps)
+        }
+    }
+
+    /// Builds up basic compilation arguments for this target.
+    public func basicArguments() -> [String] {
+        var args = [String]()
+        args += buildParameters.toolchain.clangPlatformArgs
+        args += buildParameters.flags.cCompilerFlags
+        args += optimizationArguments
+        args += ["-fobjc-arc", "-fmodules", "-fmodule-name=" + module.c99name]
+        args += ["-I", module.includeDir.asString]
+        args += additionalFlags
+        args += moduleCacheArgs
+        return args
+    }
+
+    /// Optimization arguments according to the build configuration.
+    private var optimizationArguments: [String] {
+        switch buildParameters.configuration {
+        case .debug:
+            return ["-g", "-O0"]
```

Comment:
Eventually we'll need to parameterize this, but I think for now it's fine to spell them out literally here (for one thing, we'd need to decide what to do about build parameterization first).